### ISSUE TITLE
Handling scheme-less URIs in $id field

### ISSUE THREAD

**Issue by harrel56** [harrel56/json-schema#67]
Right now if a schema has `"$id": "a/b/c"` it will get registered exactly under this id (`a/b/c`) which makes it impossible to absolutely refer to it with `$ref`. Maybe this should be registered as e.g. `https://harrel.dev/a/b/c`?

**Comment by kool79** [comment]
@harrel56 
For $id **_in subschemas_** it  should resolve it against the base URI, which is usually defined in $id at the shema root.

In case if $id **_at the root_** is also relative URI, than it should be resolved against the retrieval uri (The URI used to fetch a schema), see here for more details, with the exactly this example: https://json-schema.org/understanding-json-schema/structuring#dollarid

But if shcema is provided from source code (as a text or json element) without the base or retrieval uri, then, I think, you can assign your custom base uri (https://harrel.dev). It will be good to provide an option for library users to define the base uri when they provide schema from a code. 
 p.s. Actually the library already have API.

**Comment by harrel56** [comment]
> For $id in subschemas it should resolve it against the base URI, which is usually defined in $id at the shema root.

Yep, I believe now it works exactly like this. Actually I had in mind only cases with root schema, could be more specific about it.

> In case if $id at the root is also relative URI, than it should be resolved against the retrieval uri (The URI used to fetch a schema), see here for more details, with the exactly this example: https://json-schema.org/understanding-json-schema/structuring#dollarid

And this functionality seems to be missing - resolving `$id` against retrieval URI or some kind of base URI. Probably gonna need to investigate it further and see what exactly needs to be done here.

> p.s. Actually the library already have an API defined above and when register schema you always have the baseURI (user-defined or generated). So it can be used to resolve the relative IDs at the root

Not sure what do you mean exactly? Schemas register by `registerSchema()` call still get additional ID with which it can be referred to (regardless if `$id` is present or not) e.g. `https://harrel.dev/cfc08034`. So theoretically you could refer to such schema - is that what you meant?

>It will be good to provide an option for library users to define the base uri when they provide schema from a code.

Hmm, this indeed seems like a nice feature to have - I'll think about this when other concerns with `$id` and generally schema resolution & referencing are solved.

BTW Thanks for your input, really appreciate it :)

**Comment by kool79** [comment]
> Not sure what do you mean exactly?

Sorry, I was not clear here. Actually, my words were partially incorrect because I made some wrong assumptions.  Will try now to provide my inputs after additional investigation:
Currently we have 2 kinds of the `registerSchema` methods:
 - First accepts only the source of schema (as a string or JsonNode or schemaProviderNode)
 - Second accepts the source of schema and URI parameter.
Internally they are joined into one form,  where you use the `generateSchemaUri()` in case if user did not provide URI. So you always have some URI for every registered schema.

But currently you use this URI only as backup, i.e. when the provided schema does not contain any $id.
I think you need to use this URI like an 'retrieval uri" (below I'll use the 'retrievalUri' name for it). It means that you should resolve the $id from schema against this URI, something like: 
```
 URI  resultSchemaId = retrievalUri.resolve( id==null ? "" : id);
```
 As a result we'll have:
 - when schema does not contain any $id (i.e. it is empty) than the resultSchemaId will be equal to retrievalUri  (the same behavior as now)
 - when schema contains a absolute uri in the $id,  than the resultSchemaId will be equal to the $id (same behavior as now)
 - when schema contains a relative uri in the $id,  than resultSchemaId will be equal to the $id combined with the retrievalUri. **_This part is missed in library. But I think It should be implemented_** It will allow to correctly register different existing schemas, which have a relative URI inside $id, under the correct absolute uris. For example, when we have a schemas hierarchy in file system over different folders, but every schema contains $id with the current filename only ($id='./thisfile.json').

Note that in the last case the resultSchemaId will be absolute only when the retrievalUri is absolute. But it is up to user to provide the correct retrievalUri.  We have 2 options here: 
    - throw exception when user tries to register schema where absolute uri is not resolved (both retrievalUri and $id are relative)
    - introduce some 'top-level' absoluteDefaultUri and calculate the final schama uri as: `absoluteDefaultUri .resolve(retrievalUri.resolve(id))`
 
The 'top-level' uri may be different. It can be something with custom uri-scheme (`URI.create("harrel-schema:/")`) or we can ask user to set it:
    
```java
URI absoluteDefaultUri = URI.create("harrel-schema:/");   // with custom scheme, can be library-default
or URI absoluteDefaultUri = URI.create("file:/d:/my-schemas/");
or URI absoluteDefaultUri = URI.create("https://mySite.com");

var validator = new ValidatorFactory()
		.withTopLevelUri(absoluteDefaultUri)  // should throw if uri.isAbsolute() == false
		.createValidator();
validator.registerSchema(rootSchema);  // use absoluteDefaultUri to resolve relative $ids. Note: dont apply https:
validator.registerSchema(URI.create("./types/"), subType1Shema);  // use absoluteDefaultUri + /types/  + relative $id
validator.registerSchema(URI.create("./types/"), subType2Shema);  // use absoluteDefaultUri + /types/  + relative $id
validator.registerSchema(URI.create("resource:/globalTypes"), globalSubTypeShema);  // use resource:/globalTypes
```

Note that `validator.registerSchema(rootSchema)` should be changed and to not generate unique uri (https://harrel.dev/xxxxxxxx") for schema now.  _(IMHO the functionality with unique uri for every loaded schema should have more obvious method name, like `validator.registerSchemaWithUniqueBaseUri()`. Because this will not work correctly with relative $id-s like "./a/" or "a/" when we'll have cross-refrence between loaded schemas.)_

p.s. method and variable names in my examples aren not perfect and subject to change ))

**Comment by harrel56** [comment]
Generally it makes sense and I agree with the most of it :)

> introduce some 'top-level' absoluteDefaultUri and calculate the final schama uri as:

Definitely was thinking about this option.

> The 'top-level' uri may be different. It can be something with custom uri-scheme

Also in code snippet you mention to "not apply https". Why not? Why using custom uri-scheme could be better? As per my understanding this shouldn't really matter - if user doesn't set base URI explicitly it means that he doesn't really care about the URIs of his schemas, so potentially they could be anything.

> Note that validator.registerSchema(rootSchema) should be changed and to not generate unique uri (https://harrel.dev/xxxxxxxx") for schema now.

I think this is the main difference in our thinking. From my perspective generating unique id for each schema is must-have feature (not yet sure how to integrate this generation with setting custom base URI by user though). I want to ensure that user can call `registerSchema` several times and that it will register each schema correctly, like:
```java
var uri1 = validator.registerSchema(schema1);
var uri2 = validator.registerSchema(schema2);
var uri3 = validator.registerSchema(schema3);

var result1 = validator.validate(uri1, instance1);
var result2 = validator.validate(uri2, instance2);
var result3 = validator.validate(uri3, instance3);
```

> Because this will not work correctly with relative $id-s like "./a/" or "a/" when we'll have cross-refrence between loaded schemas.)

Maybe I'm not seeing something there but it seem to be kind of problematic in either way (generating URI or not). So maybe I'll do a step by step of my reasoning:

schema1
```json
{
  "$id": "a/b/c",
  "$ref": "x"
}
```
schema2
```json
{
  "$id": "x"
}
```

Let's assume the base URI is `https://harrel.dev`

1. `validator.validate("https://harrel.dev/a/b/c", "{}")` - no schema found, fallbacking to `SchemaResolver`
2. `SchemaResolver` gets called with "https://harrel.dev/a/b/c" uri -> resolves it to schema1
3. Now the schema needs to be registered under some URI - do we just use retrieval URI ("https://harrel.dev/a/b/c") or also try to do something with `$id` value?
4. Resolving `$ref` to "https://harrel.dev/a/b/x" - no schema found, fallbacking to `SchemaResolver`
5. `SchemaResolver` gets called with "https://harrel.dev/a/b/x" uri -> resolves it to schema2
6. Same thing as in point **3** - do we register this schema under  "https://harrel.dev/a/b/x" or also somehow resolve it against `$id` value?

I feel like it would behave the same if the base URI included generated part (e.g. `https://harrel.dev/123123`)?

Also just to mention it - as it can be seen in steps above: data for which SchemaResolver is called is going to change, so this unfortunately looks like **backward compatibility breaking change**.

**Comment by harrel56** [comment]
Will be available with the next release
